### DISCUSSION TITLE
Implicit length limit for string passed to ui.log.push()?

### DISCUSSION THREAD

**Discussion by seahawk1986** [zauberzeug/nicegui#866] [Q&A] [answered]
### Question

I found some odd behavior when building a simple log viewer (using https://nicegui.io/documentation/log) that relies on [python-systemd](https://github.com/systemd/python-systemd) to access the system's journal.

Iterating over each entry and pushing it to the ui.log instance works, but is too slow for a large file (you can see individual lines being added, which can take a while for several thousand lines) - being able to populate the log viewer with an initial data set would be helpful.
<details><summary>Naive approach</summary>
<p>

```python
from nicegui import ui
from systemd import journal


@ui.page('/')
async def logviewer():
    with ui.card().tight().classes('w-full'):
        with ui.card_section().classes('w-full'):
            ui.label("Syslog")

            j = journal.Reader()
            j.this_machine()
            j.this_boot()

            logviewer = ui.log(max_lines=10_000).classes('w-full').style("height: 75vh")

            def read_entries(logviewer, journal_reader):
                lines = (f"{entry.get('__REALTIME_TIMESTAMP')} {entry.get('MESSAGE')}" for entry in j)
                for l in lines:
                    logviewer.push(l)
            
            ui.timer(0.1, lambda: read_entries(logviewer=logviewer, journal_reader=j))


ui.run()
```
</p>
</details> 

Sending all data at once fails silently (no errors on the python side) if the number of log entries/characters in the string sent is already large enough (the limit seems to be somewhere around 217527 characters on my machine):
<details><summary>Send all existing log entries at once</summary>
<p>

```python
from nicegui import ui
from systemd import journal


@ui.page('/')
async def logviewer():
    with ui.card().tight().classes('w-full'):
        with ui.card_section().classes('w-full'):
            ui.label("Syslog")

            j = journal.Reader()
            j.this_machine()
            j.this_boot()

            logviewer = ui.log(max_lines=10_000).classes('w-full').style("height: 75vh")

            def read_entries(logviewer, journal_reader):
                lines = [f"{entry.get('__REALTIME_TIMESTAMP')} {entry.get('MESSAGE')}" for entry in j]
                if lines:
                    s = "\n".join(lines)
                    logviewer.push(s)
            
            ui.timer(0.1, lambda: read_entries(logviewer=logviewer, journal_reader=j))


ui.run()
```
</p>
</details> 

My workaround is to slice the log lines into chunks of 1000 (which should leave enough margin for longer lines) before converting them to a string:

<details><summary>Sending chunks of data</summary>
<p>

```python
from itertools import islice
from nicegui import ui
from systemd import journal

def chunk(it, size):
    it = iter(it)
    return iter(lambda: tuple(islice(it, size)), ())


@ui.page('/')
async def logviewer():
    with ui.card().tight().classes('w-full'):
        with ui.card_section().classes('w-full'):
            ui.label("Syslog")

            j = journal.Reader()
            j.this_machine()
            j.this_boot()

            logviewer = ui.log(max_lines=10_000).classes('w-full').style("height: 75vh")

            def read_entries(logviewer, journal_reader):
                lines = (f"{entry.get('__REALTIME_TIMESTAMP')} {entry.get('MESSAGE')}" for entry in j)
                for c in chunk(lines, 1_000):
                    s = "\n".join(c)
                    logviewer.push(s)
            
            ui.timer(0.1, lambda: read_entries(logviewer=logviewer, journal_reader=j))


ui.run()

```
</p>
</details>

In the console of the browser's developers tools there is also an uncaught URIError with the last two code examples:

```
 Uncaught (in promise) URIError: URI malformed
    at decodeURIComponent (<anonymous>)
    at Proxy.push (log:15:23)
    at a.<anonymous> (logs:178:81)
    at L.emit (socket.io.min.js:6:7123)
    at a.value (socket.io.min.js:6:36453)
    at a.value (socket.io.min.js:6:36157)
    at a.value (socket.io.min.js:6:35811)
    at L.emit (socket.io.min.js:6:7123)
    at socket.io.min.js:6:42922
```

**Comment by phoskee** [comment]
On discord a user had a similar problem, it seems that escape characters are being misinterpreted.
I am working on a fix, I can't run your code, but if you provide me with the part of the text that fails I can better understand and see if the fix makes a difference

**Comment by falkoschindler** [comment]
Thanks for bringing this up, @seahawk1986!
We experienced similar issues, e.g. with `ui.upload` (https://github.com/zauberzeug/nicegui/issues/685).
It might have something to do with the websocket package size. But we're not sure. Any suggestions are very welcome.

**Comment by seahawk1986** [comment]
Part of the Problem are percentage signs followed by other characters, that can't be interpreted by `decodeURIComponent` properly - e.g. I can provoke the URIError if I create an input field 
```python
from nicegui import ui

line = ui.input(label="Log input", placeholder="log line")
log = ui.log(max_lines=10).classes("w-full h-20")

def log_line():
    log.push(line.value)

ui.button("Send input to log", on_click=log_line)

ui.run()
```
And then fill the input field with `%foo` and send it to the log. The line doesn't appear in the log viewer and you get the URIError in the console of the developer tools of your browser.

If I understand it correctly https://github.com/zauberzeug/nicegui/blob/80dfbfe9d4002efbb3e6ebe0755255c5880aa3b5/nicegui/elements/log.py#L30 passes the pushed string via https://github.com/zauberzeug/nicegui/blob/80dfbfe9d4002efbb3e6ebe0755255c5880aa3b5/nicegui/element.py#L250 into the message queue (https://github.com/zauberzeug/nicegui/blob/80dfbfe9d4002efbb3e6ebe0755255c5880aa3b5/nicegui/outbox.py#L20), then it gets send via the sio (an AsyncServer from globals.py) - all without any escaping - I haven't figured out yet how https://github.com/zauberzeug/nicegui/blob/80dfbfe9d4002efbb3e6ebe0755255c5880aa3b5/nicegui/elements/log.js#L14 is called and why `line` it is an URIComponent that needs to be decoded.

**Reply by rodja** [reply, marked as answer]
Thanks for the analysis. We are just releasing 1.2.12 which improves the escaping in `ui.log`. See https://github.com/zauberzeug/nicegui/pull/865 for details.

**Comment by seahawk1986** [comment]
Unfortunately the problem I mentioned in https://github.com/zauberzeug/nicegui/discussions/866#discussioncomment-5813313 is still there with nicegui 1.2.12:
```python
from nicegui import ui
import nicegui

ui.label(f"{nicegui.__version__=}")

line = ui.input(label="Log input", placeholder="log line")
log = ui.log(max_lines=10).classes("w-full h-20")


def log_line():
    log.push(line.value)


ui.button("Log input", on_click=log_line)

ui.run()
```


![Demo](https://i.imgur.com/PTuW3M0.gif)

**Reply by falkoschindler** [reply]
Oh, that's unfortunate! I wondered how this could have passed our test. But it turns out the push works during construction time, but not dynamically when the client is already connected:
```py
log = ui.log()
log.push('50%')  # works
ui.button('push 100%', on_click=lambda: log.push('100%'))  # fails
```
I created a new issue: #899.
